### PR TITLE
Add Split-18 - num_outputs attribute, support for uneven tensor splitting

### PR DESCRIPTION
**Pull request by p-wysocki**
### Description
- new `num_outputs` attribute
- support for splitting tensors into uneven chunks added

### Motivation and Context
- solves https://github.com/onnx/onnx/issues/4188
- solves https://github.com/onnx/onnx/issues/4438

Previously discussed in https://github.com/onnx/onnx/pull/4389

### CODE REVIEW

File: onnx/backend/test/case/node/split.py [hunk 1]
```diff
@@ -113,7 +117,165 @@ def export_default_values() -> None:
...
+            node,
+            inputs=[input, split],
+            outputs=[y for y in expected_outputs],
+            name="test_split_variable_parts_default_axis_opset13",
+            opset_imports=[onnx.helper.make_opsetid("", 13)],
+        )
+
+    @staticmethod
+    def export_zero_size_splits_opset13() -> None:
+        input = np.array([]).astype(np.float32)
```
Comment (gramalingam) [inline onnx/backend/test/case/node/split.py:143]: Can you clarify the input/output shapes explicitly to help the reader? I assume that the shapes are [0] (a 1-dimensional tensor with dimension-size=0).

**Comment by gramalingam** [comment]
LGTM, thanks! Just one minor comment above to clarify shapes for the zero-size tensor.

**Review by jcwchen — COMMENTED** [review]
Perhaps you can try to cherry-pick this PR: https://github.com/onnx/onnx/pull/4202 and see whether it helps to show the exact errors without seg fault.

If that PR is indeed helpful, I would say we should forward that PR for better debugging.

**Review by jcwchen — COMMENTED** [review]
Sorry, I think my PR https://github.com/onnx/onnx/pull/4202 now needs a few updates for being compatible with https://github.com/onnx/onnx/pull/4470. We can try another way: to faster test the unknown failure in your local, perhaps you can try:
```
python setup.py install  # build ONNX with your PR
python onnx/backend/test/cmd_tools.py generate-data
git diff --exit-code --diff-filter=ADR -- . :!onnx/onnx-data.proto 
```
I suspect the number of input_*.pb or output_*.pb might be incorrect in your PR. If that is true, the commands above should show some files. If not, there might be other issues I don't know in input_*.pb or output_*.pb.

**Comment by p-wysocki** [comment]
@jcwchen I ran the commands, they did not show any files. However, after removing `test_driver.cc` changes the CI started passing. I have no idea why it's passing now, maybe the issue was sporadic or has been fixed on master in the meantime. Thanks for the help!

**Review by jcwchen — CHANGES_REQUESTED** [review]
I might figure out why previous CI has failed: some added test model files are not needed. I don't see where they are generated. For instance, `test_uneven_split` and `test_uneven_split_opset18`. I would suggest you clean all of added test files and then use `python onnx/backend/test/cmd_tools.py generate-data` to generated them again. Since related test was just removed due to recent ONNXIFI deprecation, I will add a similar check in CI to prevent future issues like it.

Another suggestion would be making test files naming convention consistent. For names of Split related tests, please use `test_split_XXX_XXX`. IMO, it will be helpful for users to search related test files.

**Comment by p-wysocki** [comment]
@jcwchen I applied your suggestions, thanks! For some reason I assumed that script that generates data also removes the old files. I also approved your PR introducing a CI check for that, it sounds helpful. Test names also have been aligned.

**Review by jcwchen — COMMENTED** [review]
Thank you for the updates! Once you have updated `TestCoverage.md` by the updated test scripts , I think this PR is ready to go.

**Review by jcwchen — APPROVED** [review]

**Comment by gramalingam** [comment]
> I might figure out why previous CI has failed: some added test model files are not needed. I don't see where they are generated. For instance, `test_uneven_split` and `test_uneven_split_opset18`. I would suggest you clean all of added test files and then use `python onnx/backend/test/cmd_tools.py generate-data` to generated them again. Since related test was just removed due to recent ONNXIFI deprecation, I will add a similar check in CI to prevent future issues like it.
> 
> Another suggestion would be making test files naming convention consistent. For names of Split related tests, please use `test_split_XXX_XXX`. IMO, it will be helpful for users to search related test files.

I think it would be helpful to update the `AddNewOp` document (may be as a helpful Troubleshooting tip at the end) to remind people to delete the old folder before generating test-cases. (If the CI can catch it, that would be great too.)

File: onnx/defs/tensor/defs.cc [hunk 2]
```diff
@@ -603,15 +603,16 @@ ONNX_OPERATOR_SET_SCHEMA(
...
         }));
 
-static const char* Split_ver13_doc =
-    R"DOC(Split a tensor into a list of tensors, along the specified
-'axis'. Lengths of the parts can be specified using input 'split'.
-Otherwise, the tensor is split to equal sized parts.
+static const char* Split_ver18_doc =
+    R"DOC(Split a tensor into a list of tensors, along the specified 'axis'.
+Either input 'split' or the attribute 'num_outputs' should be specified, but not both.
+If the attribute 'num_outputs' is specified, then the tensor is split into equal sized parts.
```
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc:609]: Need to update above line by adding the line from below ``If the tensor is not evenly splittable, the last chunk will be smaller.``

**Comment by jcwchen** [comment]
> I think it would be helpful to update the AddNewOp document (may be as a helpful Troubleshooting tip at the end) to remind people to delete the old folder before generating test-cases. (If the CI can catch it, that would be great too.)

Sounds good. I will add sentence about it in AddNewOp.md in this PR as well: https://github.com/onnx/onnx/pull/4514. Also note that that PR will let CI to catch this kind of issue. Perhaps we can forward that PR first and have that verify this PR.

> Or, may be the test-case generator should delete the folder? (Of course, it could be a bit risky in case the user has some other file there ... may be warn the user if there is some extra files there perhaps.)

Yes, but it seems possible that users would like to keep existing ones for some testing purposes? Therefore, to keep behavior consistent, I introduced --clean flag for generate_data for CI use (by default it is False so it won't clean original node directory automatically).

File: onnx/defs/tensor/defs.cc (hunk 2 above)
Comment (p-wysocki) [reply onnx/defs/tensor/defs.cc:609]: Good catch, done.

**Review by gramalingam — APPROVED** [review]

**Review by jcwchen — APPROVED** [review]
It passed the enhanced CI 💯  Thank you @p-wysocki for this PR and waiting! I will merge it soon.

**Pull request merged**